src/libos/crates/sworn-disk/src/index/mem_table.rs

```diff
@@ -0,0 +1,332 @@
+//! MemTable of lsm tree.
+use crate::data::CacheState;
+use crate::prelude::*;
+use crate::util::RangeQueryCtx;
+use crate::{Checkpoint, Record};
+
+use std::collections::BTreeMap;
+use std::ops::Bound::Included;
+
+/// MemTable. Define common memtable ability.
+pub trait MemTable<K: Into<LBA>, V: Clone> {
+    /// Get an entry with a key.
+    fn get(&self, key: &K) -> Option<V>;
+
+    /// Put a new entry(kv) into memtable.
+    fn put(&mut self, key: K, value: V) -> Option<HBA>;
+
+    /// Get entries with a range of key.
+    fn get_range(&self, first: &K, last: &K) -> Vec<V>;
+
+    /// get all records from memtable.
+    fn get_all_records(&self) -> Vec<Record>;
+
+    /// Clear memtable to empty.
+    fn clear(&mut self);
+
+    /// Capacity (max size).
+    fn capacity(&self) -> usize;
+
+    /// Current size.
+    fn size(&self) -> usize;
+
+    /// Whether memtable becomes full (size reach to capacity)
+    fn is_full(&self) -> bool;
+
+    /// Whether memtable contains a entry with a key.
+    fn contains(&self, key: &K) -> bool;
+}
+
+/// `BTreeMemTable: MemTable<LBA, MemRecord>` uses `BTreeMap` to track records.
+pub struct BTreeMemTable {
+    // TODO: Use RBTree
+    records: BTreeMap<LBA, MemRecord>,
+    size: usize,
+    capacity: usize,
+}
+
+/// `Record` in memory.
+#[derive(Clone, Debug)]
+pub enum MemRecord {
+    /// A regular record
+    Regular(Record),
+    /// A negative record, only its lba is meaningful
+    Negative(LBA),
+    /// Same lba, double records (one is regular, the other is negative)
+    RegularAndNegative(Record),
+}
+
+impl MemTable<LBA, MemRecord> for BTreeMemTable {
+    fn get(&self, key: &LBA) -> Option<MemRecord> {
+        self.records.get(key).map(|record| record.clone())
+    }
+
+    fn put(&mut self, key: LBA, value: MemRecord) -> Option<HBA> {
+        let mut reclaimed_hba = None;
+
+        let existed_record = self.records.get(&key);
+        match (&value, &existed_record) {
+            (MemRecord::Regular(_), Some(MemRecord::Regular(r2))) => {
+                self.size -= 1;
+                let _ = reclaimed_hba.insert(r2.hba);
+                self.records.insert(key, value.clone());
+            }
+            (MemRecord::Regular(r1), Some(MemRecord::Negative(_))) => {
+                self.records
+                    .insert(key, MemRecord::RegularAndNegative(r1.clone()));
+            }
+            (MemRecord::Regular(r1), Some(MemRecord::RegularAndNegative(r2))) => {
+                self.size -= 1;
+                let _ = reclaimed_hba.insert(r2.hba);
+                self.records
+                    .insert(key, MemRecord::RegularAndNegative(r1.clone()));
+            }
+            (MemRecord::Negative(_), Some(MemRecord::Regular(_))) => {
```

Comment:
Under current implementation, A `Negative` is put into memtable only when GC started and check a HBA is still valid (no newer data).

`Regular` and `Negative` is easy to understand. But `RegularAndNegative` here always means the regular one is newer coming than the negative one.

We only reclaim a HBA when `Regular` meet a `Regular`.

When a `Negative` arrives, all older ones can be deprecated (they are already reclaimed and are now processed by GC)

The `Negative` ones cannot be removed directly in memtable since they are needed in Compaction.

Workflow in memtable:

| newer    | older              |                                                           |
| -------- | ------------------ | --------------------------------------------------------- |
| Regular  | Regular            | reclaim immediately and insert newer one                  |
| Regular  | Negative           | replace older one to `RegularAndNegative`                 |
| Regular  | RegularAndNegative | reclaim immediately                                       |
| Negative | Regular            | replace older one to newer one (only one `Negative` left) |
| Negative | RegularAndNegative | replace older one to newer one (only one `Negative` left) |
| _        | None               | insert newer one                                          |
| _        | _                  | should not happen                                         |

And I should add this to comment.